python3Packages.catboost: 1.0.5 -> 1.2.2; build with cmake

natsukium opened the pull request:
###### Description of changes

Diff: https://github.com/catboost/catboost/compare/v1.0.5...v1.2.2
Changelog: https://github.com/catboost/catboost/releases/tag/v1.2.2

> CatBoost's build system has been switched from Ya Make (Yandex's build system) to [CMake](https://cmake.org/). This means more transparency in the build process and more familiar tools for Open Source developers.
For now it is possible to build CatBoost for:
Linux on x86-64 with or without CUDA
Linux on aarch64 with or without CUDA
macOS on x86-64 and arm64, including creating universal binaries
Windows on x86-64 with or without CUDA
Android (only model applier) on [All supported ABIs](https://developer.android.com/ndk/guides/abis).

Following @nviets suggestion in #208317, I split the cli application and shared libraries into separate packages.
I have successfully built cli and python-module.

The rest of the tasks are
- [x] build on other than `x86_64-linux`
  - [x] `aarch64-linux`
  - [x] `x86_64-darwin`
  - [x] `aarch64-darwin` 
- [x] build with cuda

Skip in this PR
- [ ] ~~build the R libraries~~

I'm not familiar with JVM.
- [ ] ~~build the JVM libraries~~
- [ ] ~~build the spark libraries~~

cc @veprbl @PlushBeaver

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

natsukium committed:
catboost: init at 1.2.2

natsukium committed:
python310Packages.catboost: 1.0.5 -> 1.2.2

Diff: https://github.com/catboost/catboost/compare/v1.0.5...v1.2.2

Changelog: https://github.com/catboost/catboost/releases/tag/v1.2.2

nviets commented:
Exciting work, @natsukium! That conan part really got me when I gave this a try - cool to see you could just drop it.

The R package should be handled similarly to [xgboost](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/xgboost/default.nix#L66), since catboost isn't in CRAN and we'd like the CUDA support. The R support can be handled in a separate commit, though, so feel free to merge when you're ready.

nviets commented:
Result of `nixpkgs-review pr 235226` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.shap</li>
    <li>python310Packages.shap.dist</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python311Packages.catboost</li>
    <li>python311Packages.catboost.dist</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>catboost</li>
    <li>catboost.dev</li>
    <li>python310Packages.catboost</li>
    <li>python310Packages.catboost.dist</li>
  </ul>
</details>

nviets commented:
###### Failed derivations

- [ ] python3.11-catboost-1.2.drv (derivation hash: 93qc2ic1i8hi7vba33psl732i5xinpxa)
  - https://gist.github.com/nviets/7f0bad210607d9082a73e773d2a7d576

natsukium commented:
Thanks for trying.
Looking at the logs, it looks like we need to match the python version when compiling libcatboost.
I will work on it this weekend.

nviets commented:
Result of `nixpkgs-review pr 235226 --extra-nixpkgs-config '{ cudaCapabilities = [ "8.6" ]; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.shap</li>
    <li>python310Packages.shap.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>catboost</li>
    <li>catboost.dev</li>
    <li>python310Packages.catboost</li>
    <li>python310Packages.catboost.dist</li>
    <li>python311Packages.catboost</li>
    <li>python311Packages.catboost.dist</li>
  </ul>
</details>

nviets commented:
@zhuk66 - FYI

natsukium commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,111 @@
+    zlib
+  ] ++ lib.optionals stdenv.isDarwin [
+    libiconv
+  ] ++ lib.optionals cudaSupport [
+    (cudatoolkit.override { backendStdenv = stdenv; })
+  ];
+
+  env = {
+    CUDAHOSTCXX = "${stdenv.cc}/bin/clang";
```
@NixOS/cuda-maintainers 
In case `backendStdenv = clangStdenv`, it needs to set `CUDAHOSTCXX = ${stdenv.cc}/bin/clang` explicitly.
This is because `-DCUDA_HOST_COMPILER=${backendStdenv.cc}/bin` is set here.
https://github.com/NixOS/nixpkgs/blob/3ae20aa58a6c0d1ca95c9b11f59a2d12eebc511f/pkgs/development/compilers/cudatoolkit/common.nix#L272-L276
Since nvcc looks for gcc by default, it tries to find `clang/bin/gcc` without the CUDAHOSTCXX.
Is it worth setting `-DCUDA_HOST_COMPILER=${lib.getExe backendStdenv.cc.cc}` in cudatoolkit?

SomeoneSerge replied:
Hi! Could you give some more intro on why this needs to be built with clang?

The reason the `backendStdenv` attribute is there is because different cudatoolkit versions support different backend compiler releases (e.g. nvcc from cudatoolkit 11.7 will simply error out when used with gcc12). At the moment we only maintain a table with nvcc-gcc versions compatibility, but in principle one could add llvm as well. The `(cudatoolkit.override { backendStdenv = clangStdenv; })` might work but only accidentally and until nixpkgs bumps the default llvm major version.

> `-DCUDA_HOST_COMPILER=${lib.getExe backendStdenv.cc.cc}`

There are two issues with this line:
1. Both the legacy `CUDA_HOST_COMPILER` and the `CMAKE_CUDA_HOST_COMPILER` map into nvcc's `-ccbin`, which expects a directory. Iirc it'd use `${ccbin}/cc` or `${ccbin}/cxx` as the actual compiler. I don't remember exactly, but I think `CUDAHOSTCXX` works the same
2. The un-wrapped `backendStdenv.cc.cc` (opposed to `backendStdenv.cc`) is something you probably don't want to use. E.g. `cudaPackages_11.backendStdenv.cc.cc` is gcc11 and it links its artifacts to gcc11's libstdc++, which won't work with the rest of nixpkgs. Similarly, if you were using clang, you'd need to wrap it so that it would link gcc's `libstdc++` instead of llvm's `libcxx`. I'm sorry about this mess, there's an open issue tracking this here: https://github.com/NixOS/nixpkgs/issues/226165

SomeoneSerge replied:
Also, you probably want to use `cudaPackages.cuda_{nvcc,cudart,..}` instead of `cudatoolkit`

natsukium replied:
@SomeoneSerge san, Thanks for the clarification.
I need clang12+ to build this catboost and can't build it with GCC.
ref: https://catboost.ai/en/docs/installation/build-environment-setup-for-cmake#compilers,-linkers-and-related-tools
I understand the role of backendStdenv, and even if accidentally, I should fix the clang version here.

> Both the legacy CUDA_HOST_COMPILER and the CMAKE_CUDA_HOST_COMPILER map into nvcc's -ccbin, which expects a directory. Iirc it'd use ${ccbin}/cc or ${ccbin}/cxx as the actual compiler. I don't remember exactly, but I think CUDAHOSTCXX works the same

Yes, `-ccbin` expects a directory, but the [documentation](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#compiler-bindir-directory-ccbin) says it is possible to specify an executable.
The `backendStdenv=clangStdenv` case is a problem because it tries to run `${ccbin}/gcc` instead of `${ccbin}/cc` as mentioned above.
In fact, building this PR change without CUDAHOSTCXX fails.

> The un-wrapped backendStdenv.cc.cc (opposed to backendStdenv.cc) is something you probably don't want to use. E.g. cudaPackages_11.backendStdenv.cc.cc is gcc11 and it links its artifacts to gcc11's libstdc++, which won't work with the rest of nixpkgs. Similarly, if you were using clang, you'd need to wrap it so that it would link gcc's libstdc++ instead of llvm's libcxx. I'm sorry about this mess, there's an open issue tracking this here: cudaPackages: extend cc-wrapper and rewrite cudaStdenv to properly solve libstdc++ issues 

Thanks for letting me know about the cc-wrapper issue.
I will change it to link libstdc++.

SomeoneSerge replied:
Btw, if you'd like to prevent cudatoolkit/cuda_nvcc from exporting CUDAHOSTCXX&c, you should be able to use `dontSetupCUDAToolkitCompilers = true` once https://github.com/NixOS/nixpkgs/pull/233581/files#diff-412326f1eb2e17a306d14d70da7fe2423b4f244aa3c9c81d13f1bb1cf3be084bR20 is merged

SomeoneSerge replied:
Hello again! You mentioned earlier that

> [Since] nvcc looks for gcc by default [...].

I may be now facing a similar issue with cross-compilation, so I was wondering if you really meant nvcc (rather than cmake, or even findcuda/findcudatoolkit specifically), and if you've got a reference documenting this behaviour?

natsukium replied:
Sorry, I can't answer this question for sure, but I guess it is the behavior of nvcc.
Looking at CMake's FindCUDAToolkit and other CUDA-related sources, it does not seem to refer specifically to gcc.
I can't find any documentation on this behavior, but JetBrain's CLion page says the following for reference.
https://www.jetbrains.com/help/clion/cuda-projects.html#custom-host-compiler
> For compiling host code, NVCC calls the system's default C++ compiler (gcc/g++ on Linux and cl.exe on Windows).

natsukium committed:
catboost: add natsukium to maintainers

natsukium commented:
Now that the cleanup is complete, it is ready for review.

SomeoneSerge commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,111 @@
+, yasm
+, zlib
+, cudaSupport ? config.cudaSupport or false
+, cudatoolkit
```
This way of consuming cuda is obsolete: `cudatoolkit` is an alias to `cudaPackages.cudatoolkit`. Attributes outside `cudaPackages` consume the entire `cudaPackages` scope and then access its individual elements, such as `cuda_nvcc` or `cudatoolkit`: https://nixos.org/manual/nixpkgs/unstable/#cuda

natsukium commented:
I've changed cudatoolkit to cuda_{nvcc,cudart,ccl}.
I've also fixed clangStdenv to clang12Stdenv to prevent accidental breakage with the clang update.
It looks like I didn't need to change backendStdenv.

SuperSandro2000 commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,114 @@
+
+  env = {
+    CUDAHOSTCXX = "${stdenv.cc}/bin/cc";
+    NIX_CFLAGS_LINK = lib.optionalString stdenv.isLinux "-fuse-ld=lld";
```
This will result in the env being set to empty, not sure if this breaks something

natsukium replied:
It should not be a problem.
I often see this in other derivations.

SuperSandro2000 commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,114 @@
+    CUDAHOSTCXX = "${stdenv.cc}/bin/cc";
+    NIX_CFLAGS_LINK = lib.optionalString stdenv.isLinux "-fuse-ld=lld";
+    NIX_LDFLAGS = "-lc -lm";
+    # protoc uses this variable
+    LD_LIBRARY_PATH = "${lib.getLib zlib}/lib";
```
Can we use patchelf instead? LD_LIBRARY_PATH is rather dirty

natsukium replied:
I didn't know that using patchelf would solve this.
I added an autoPatchelfHook. Thanks.

SomeoneSerge replied:
In the hindsight, I'm a bit confused: how exactly does `protoc` use LD_LIBRARY_PATH, and what exactly is it that autoPatchelfHook ends up patchelfing? How does the bulid fail when both autopatchelf and ld_library_path are omitted?

natsukium replied:
The build should have failed because protoc could not find zlib. 
I don't know if it was due to a version bump or my mistake, but this hack was no longer needed.

SuperSandro2000 commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,114 @@
+      --replace "\''${YASM_BIN}" "${lib.getExe yasm}"
+  '';
+
+  outputs = [
+    # binary and libraries e.g. bin/catboost, lib/catboost.so, etc.
+    "out"
+    # libraries required for Python, R, etc. support
+    # see https://catboost.ai/en/docs/installation/build-native-artifacts#targets
+    "dev"
+  ];
```
```suggestion
  outputs = [ "out" "dev" ];
```
this should follow the standard outputs outlines in the nixpkgs doc

SuperSandro2000 commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,114 @@
+  patches = [
+    ./remove-conan.patch
+    ./fix-imported-targets.patch
+  ] ++ lib.optionals (lib.versionOlder cudaPackages.cudatoolkit.version "11.8") [
```
We can't make this unconditional, can we?

natsukium replied:
Yes, we need this condition.

SuperSandro2000 commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,114 @@
+, yasm
+, zlib
+, cudaSupport ? config.cudaSupport or false
+, cudaPackages ? {}
```
Isn't that always defined?

natsukium replied:
I'm not sure if this is necessary.
I followed the [CUDA documentation](https://nixos.org/manual/nixpkgs/stable/#cuda).

SomeoneSerge replied:
Not necessary since https://github.com/NixOS/nixpkgs/pull/224068/files#diff-1977c7748af8b43b92093d2383e23e88c4df10a702578fbe885675a0956a2f1f

SuperSandro2000 commented on pkgs/development/python-modules/catboost/default.nix:
```diff
@@ -1,64 +1,63 @@
+    runHook preBuild
+
+    # these arguments must set after bdist_wheel
+    ${python.pythonForBuild.interpreter} setup.py bdist_wheel --no-widget --prebuilt-extensions-build-root-dir=${lib.getDev catboost}
```
I think there was an option added recently to make this possible in the hooks

natsukium replied:
I can't find it in `setuptoolsBuildHook`.

SuperSandro2000 commented on pkgs/development/python-modules/catboost/default.nix:
```diff
@@ -1,64 +1,63 @@
-    # _catboost.pyx.cpp:226822:19: error: use of undeclared identifier '_PyGen_Send'
-    broken = withCuda;
-  };
+  meta =
+    catboost.meta
+    // {
```
```suggestion
  meta = catboost.meta // {
```

SuperSandro2000 commented on pkgs/development/python-modules/catboost/default.nix:
```diff
@@ -1,64 +1,63 @@
+    // {
+      # ERROR: catboost-1.2-cp310-cp310-macosx_11_0_x86_64.whl is not a supported wheel on this platform.
+      # relates https://github.com/NixOS/nixpkgs/issues/229864
+      broken = stdenv.isDarwin && stdenv.isx86_64;
```
not sure if meta.platforms is more appropriate here

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1653,7 +1653,12 @@ self: super: with self; {
   catalogue = callPackage ../development/python-modules/catalogue { };
 
-  catboost = callPackage ../development/python-modules/catboost { };
+  catboost = callPackage ../development/python-modules/catboost {
+    catboost = pkgs.catboost.override {
+      enablePython = true;
+      python3 = python;
+    };
+  };
```
This is rather odd, can't we move the python build completely into pkgs.catboost?

natsukium replied:
We could perhaps build python packages in one derivation, but it would increase complexity and be inefficient with long builds each time, even if we only change the python package.

I switched to python3Packages as input instead of python3, as many other packages do.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28097,6 +28097,10 @@ with pkgs;
   cascadia-code = callPackage ../data/fonts/cascadia-code { };
 
+  catboost = callPackage ../development/libraries/catboost {
+    stdenv = clang12Stdenv;
```
clang12Stdenv is an alias, I recommend to turn them off if you are contributing a lot.

```suggestion
    inherit (llvmPackages_12) stdenv;
```

nviets commented:
[1.2.1](https://github.com/catboost/catboost/releases/tag/v1.2.1) was released. Maybe bump the version before merging?

natsukium commented on pkgs/development/libraries/catboost/remove-unsupport-cc.patch:
```diff
@@ -0,0 +1,532 @@
+   -gencode=arch=compute_80,code=sm_80
+   -gencode=arch=compute_86,code=sm_86
+-  -gencode=arch=compute_89,code=sm_89
+-  -gencode=arch=compute_90,code=sm_90
```
I used sed because there are many cmakelists to change, and it takes a long time.
Thanks.

natsukium commented:
It's ready for review again.

The main changes are.
- Update from 1.2 to 1.2.1
- Adopt Sandro's comments (with some exceptions)
- Use cuda-redist

@SuperSandro2000, @NixOS/cuda-maintainers Could you review this PR again?

SomeoneSerge commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29299,6 +29299,10 @@ with pkgs;
   cascadia-code = callPackage ../data/fonts/cascadia-code { };
 
+  catboost = callPackage ../development/libraries/catboost {
+    inherit (llvmPackages_12) stdenv;
```
A brief comment explaining why `_12` and when to bump?

SomeoneSerge commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,135 @@
+    for cmakelists in **/CMakeLists.*; do
+      sed -i "s/OpenSSL::OpenSSL/OpenSSL::SSL/g" $cmakelists
+      ${lib.optionalString (lib.versionOlder cudaVersion "11.8") ''
+        sed -i 's/-gencode=arch=compute_89,code=sm_89//g' $cmakelists
+        sed -i 's/-gencode=arch=compute_90,code=sm_90//g' $cmakelists
```
Is this dropping the support for 89, 90? Is there a way to specify the list of gencodes for which to build the kernels?

SomeoneSerge replied:
Seeing how moving `cuda_*` to `*[bB]uildInputs` has worked, and how the build logs seem to mention FindCUDAToolkit.cmake: https://gist.github.com/SomeoneSerge/395dfc734cc220aae928cd9d501acabd#file-gistfile0-txt-L52

...I think it should be possible to control the target cuda architectures by setting https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_ARCHITECTURES.html, unless upstream overrides the variable. The configurable way to choose capabilities at the moment would be to use `cudaPackages.cudaFlags`, which respect `config.cudaCapabilities` (and whatever they might get replaced with later). Examples:
https://github.com/NixOS/nixpkgs/blob/b5213d5b0f6bda0fe098ff98890599ff63dfdd5f/pkgs/development/libraries/science/math/faiss/default.nix#L103
https://github.com/NixOS/nixpkgs/blob/b5213d5b0f6bda0fe098ff98890599ff63dfdd5f/pkgs/development/libraries/science/math/magma/generic.nix#L78-L81

SomeoneSerge replied:
I just tested this in https://github.com/SomeoneSerge/nixpkgs/pull/8/commits/9780f86c4a3938c166959895cad8cdb2b3104fde and it doesn't work because upstream passes some ad hoc flags instead of using `CMAKE_CUDA_ARCHITECTURES`:

https://github.com/catboost/catboost/blob/ad3934317eb5357781e2389f1c24582c01f738ca/catboost/cuda/ctrs/CMakeLists.linux-x86_64-cuda.txt#L34-L56 https://github.com/catboost/catboost/blob/ad3934317eb5357781e2389f1c24582c01f738ca/cmake/cuda.cmake#L163-L165

I think we should contact upstream about changing that

SomeoneSerge commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,135 @@
+  ] ++ lib.optionals stdenv.isDarwin [
+    libiconv
+  ] ++ lib.optionals cudaSupport [
+    cuda-redist
```
Is symlinkJoin necessary?

natsukium replied:
It may not be necessary.
I found this approach in other packages, so I just followed the pattern.

SomeoneSerge replied:
> I found this approach in other packages, so I just followed the pattern.

The `symlinkJoin` trick is still necessary for packages that use `FindCUDA.cmake`, but the preferred way is to use `FindCUDAToolkit.cmake` and avoid `symlinkJoin`. One of the costs of `symlinkJoin` is increased runtime closure size

natsukium replied:
Thanks, that makes sense.

natsukium commented:
things done

https://github.com/NixOS/nixpkgs/compare/97135d6fa36159f04c37529ab90820a4462d37f1..7425a11b43a3f3ce8e7880f61d1ec42063230267
- use `config.cudaSupport` without `or false`
- remove `symlinkJoin`
- add a comment for using `llvmPackages_12`
- drop `autoPatchelfHook`

https://github.com/NixOS/nixpkgs/compare/7425a11b43a3f3ce8e7880f61d1ec42063230267..0b93e56237687efe526ab2a595356debee9da50c
- update 1.2.1 -> 1.2.2

SomeoneSerge commented:
Result of `nixpkgs-review pr 235226 --extra-nixpkgs-config '{ cudaCapabilities = [ "8.6" ]; cudaSupport = true; cudaForwardCompat = false; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python310Packages.shap</li>
    <li>python310Packages.shap.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>catboost</li>
    <li>catboost.dev</li>
    <li>python310Packages.catboost</li>
    <li>python310Packages.catboost.dist</li>
    <li>python311Packages.catboost</li>
    <li>python311Packages.catboost.dist</li>
  </ul>
</details>

SomeoneSerge commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,113 @@
+  ]);
+
+  env = {
+    CUDAHOSTCXX = lib.optionalString cudaSupport "${stdenv.cc}/bin/cc";
```
We set this variable in `setup-cuda-hook.sh` (propagated by cuda_nvcc in natvieBuildInputs), so this should be a no-op. Also, if upstream uses `FindCUDAToolkit` and `enable_language(CUDA)`, then `CUDAHOSTCXX` is superseded by `CMAKE_CUDA_HOST_COMPILER` (which the hook should set as well)

natsukium replied:
As reported in [the first comment](https://github.com/NixOS/nixpkgs/pull/235226#discussion_r1240595982), nvcc uses gcc instead of clang as the backendStdenv without it.
We will get the following error message if we do not set this environment variable.

```
catboost> [2892/4763] Building CUDA object library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o
catboost> FAILED: library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o 
catboost> /nix/store/mr3zd7dhyxdwr9y9gf6l74fah1rkjkji-cuda_nvcc-11.8.89/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/nix/store/848ld56pzbywlxy1pv9xi5d8gsvzb6hl-gcc-wrapper-11.4.0/bin/c++ -DCATBOOST_OPENSOURCE=yes -I/build/source -I/build/source/build -I/build/source/contrib/libs/nvidia/thrust -I/build/source/contrib/libs/nvidia/cub -I/build/source/contrib/libs/linux-headers -I/build/source/contrib/libs/linux-headers/_nf -I/build/source/contrib/libs/cxxsupp/libcxx/include -I/build/source/contrib/libs/cxxsupp/libcxxrt/include -I/build/source/contrib/libs/clang14-rt/include -I/build/source/contrib/libs/zlib/include -I/build/source/contrib/libs/double-conversion -I/build/source/contrib/libs/libc_compat/include/readpassphrase -I/build/source/contrib/libs/libc_compat/reallocarray -I/build/source/contrib/libs/libc_compat/random -isystem /nix/store/mr3zd7dhyxdwr9y9gf6l74fah1rkjkji-cuda_nvcc-11.8.89/include -D_THREAD_SAFE -D_PTHREADS -D_REENTRANT -D_LARGEFILE_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__LONG_LONG_SUPPORTED -D_GNU_SOURCE -DLIBCXX_BUILDING_LIBCXXRT -D_FILE_OFFSET_BITS=64 -D_YNDX_LIBUNWIND_ENABLE_EXCEPTION_BACKTRACE -DSSE_ENABLED=1 -DSSE3_ENABLED=1 -DSSSE3_ENABLED=1 -DSSE41_ENABLED=1 -DSSE42_ENABLED=1 -DPOPCNT_ENABLED=1 -DCX16_ENABLED=1 --compiler-options -fexceptions,-fno-common,-fcolor-diagnostics,-fdebug-default-version=4,-ffunction-sections,-fdata-sections,-Wall,-Wextra,-Wno-parentheses,-Wno-implicit-const-int-float-conversion,-Wno-unknown-warning-option,-pipe,-fuse-init-array,-m64,-mpopcnt,-mcx16,-Woverloaded-virtual,-Wimport-preprocessor-directive-pedantic,-Wno-undefined-var-template,-Wno-return-std-move,-Wno-defaulted-function-deleted,-Wno-pessimizing-move,-Wno-deprecated-anon-enum-enum-conversion,-Wno-deprecated-enum-enum-conversion,-Wno-deprecated-enum-float-conversion,-Wno-ambiguous-reversed-operator,-Wno-deprecated-volatile --expt-extended-lambda --expt-relaxed-constexpr --compiler-options -std=c++14 -DTHRUST_IGNORE_CUB_VERSION_CHECK --threads 0 -O3 -DNDEBUG -std=c++14 -Xcompiler=-fPIC -nostdinc++ -DLIBCXX_BUILDING_LIBCXXRT -D_libunwind_ -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=compute_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=compute_60 -gencode arch=compute_61,code=compute_61 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_70,code=compute_70 --ptxas-options=-v -lineinfo --use_fast_math -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -MD -MT library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o -MF library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o.d -x cu -c /build/source/library/cpp/cuda/wrappers/kernel.cu -o library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o
catboost> nvcc warning : incompatible redefinition for option 'compiler-bindir', the last value of this option was used
catboost> nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> fatal   : Could not open input file /build/tmpxft_0000501d_00000000-16_kernel.compute_35.cpp1.ii
```

SomeoneSerge commented on pkgs/development/libraries/catboost/default.nix:
```diff
@@ -0,0 +1,113 @@
+
+  postPatch = ''
+    substituteInPlace cmake/common.cmake \
+      --replace  "\''${RAGEL_BIN}" "${ragel}/bin/ragel" \
+      --replace "\''${YASM_BIN}" "${yasm}/bin/yasm"
```
Cosmetic: `lib.getExe ragel` would probably work too

natsukium replied:
Neither package has mainProgram set yet, and using getExe will cause a warning.😞

SomeoneSerge reviewed (approved):
I've only skimmed the changes this time, but I think this is ready. The failing packages in nixpkgs-review are broken on master as well (cf hydra). Thank you a lot @natsukium!

nviets commented:
@natsukium, @SomeoneSerge - are you able to run the GPU version successfully with CLI and/or Python? I started a branch on your work to add the R package, and everything is working except for the GPU. My branch is [here](https://github.com/nviets/nixpkgs/blob/catboost122R/pkgs/development/libraries/catboost/default.nix), and I am waiting for this one to merge before opening the draft PR. Mine fails on:

```
➜ ./result/bin/Rscript test.R
Loading required package: ggplot2
Loading required package: lattice
Error in catboost.train(pool_train, pool_test, params) :
  /build/source/catboost/cuda/cuda_lib/cuda_base.h:281: CUDA error 35: CUDA driver version is insufficient for CUDA runtime version
Execution halted
```

I guess my two questions are 1) am I handling the R library build correctly because I'm not sure if its supposed to reference the /build/* directory like that and 2) why is it failing on this CUDA message? I tried several versions of CUDA, which makes me think I'm missing a header file or something.

Your work looks good to go, so I don't want this question to hold up your merge. We can check it out on my follow up PR later. Thanks!

SomeoneSerge commented:
@nviets thanks, good catch, I tested with https://github.com/SomeoneSerge/nixpkgs/commit/56e2da21376c85472847e6fcfbe76d45526bea70 and got the same error: it's just `_catboost.so` trying to dlopen libcuda.so. We need to add the respective runpath: https://github.com/SomeoneSerge/nixpkgs/pull/8/commits/a290d7c49c61b51888b2f8802e120a0cacdb1675

nviets commented:
Thanks @SomeoneSerge - that did the trick for the R package too.

veprbl commented:
Result of `nixpkgs-review pr 235226` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.shap</li>
    <li>python310Packages.shap.dist</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>catboost</li>
    <li>catboost.dev</li>
    <li>python310Packages.catboost</li>
    <li>python310Packages.catboost.dist</li>
    <li>python311Packages.catboost</li>
    <li>python311Packages.catboost.dist</li>
  </ul>
</details>

veprbl reviewed (approved):
LGTM

natsukium commented:
@nviets Thanks for the report. I didn't notice that since I was working with a WSL that handled LD_LIBRARY_PATH differently.
> am I handling the R library build correctly because I'm not sure if its supposed to reference the /build/* directory like that 

All the built libraries and binaries are contained under $dev, and the directory tree follows the output location in [the official documentation](https://catboost.ai/en/docs/installation/build-native-artifacts#targets).

@SomeoneSerge Thanks for investigating the runpath issue.
Is this an issue that should be fixed in magma rather than catboost?

natsukium commented:
I'll merge this to move forward.
